### PR TITLE
PostgreSQL wire protocol implementation

### PR DESCRIPTION
**Pull request by MovElb**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Implementation of PostgreSQL wire protocol.

Detailed description / Documentation draft:
Added TCPHandler, which allows PostgreSQL clients to connect ClickHouse.

### CODE REVIEW

File: programs/server/PostgreSQLHandler.h [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+namespace CurrentMetrics
+{
+    extern const Metric PostgreSQLConnection;
+}
+
+namespace DB
+{
+
+using namespace DB::PostgreSQLProtocol;
```
Comment (vitlibar) [inline programs/server/PostgreSQLHandler.h:20]: Please don't use `using namespace`.
https://clickhouse.tech/docs/en/development/style/

File: programs/server/PostgreSQLHandler.cpp [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+                            "ClickHouse doesn't support exteneded query mechanism"),
+                        true);
+                    LOG_ERROR(log, "Client tried to access via extended query protocol");
+                    message_transport->dropMessage();
+                    break;
+                default:
+                    message_transport->send(
+                        Messaging::ErrorOrNoticeResponse(
+                            Messaging::ErrorOrNoticeResponse::ERROR,
+                            "0A000",
```
Comment (vitlibar) [inline programs/server/PostgreSQLHandler.cpp:94]: What does `0A000` mean?

File: tests/integration/test_postgresql_protocol/configs/users.xml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+ <yandex>
+     <profiles>
+         <default>
+         </default>
+     </profiles>
+
+     <users>
+         <default>
+             <password>123</password>
+             <profile>default</profile>
```
Comment (vitlibar) [inline tests/integration/test_postgresql_protocol/configs/users.xml:11]: The `<profile>` and `<quota>` lines are not necessary and could be deleted along with the `<profiles>` and `<quotas>` sections.

File: src/Processors/Formats/Impl/PostgreSQLOutputFormat.h [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+#include <Processors/Formats/IRowOutputFormat.h>
+#include <Core/Block.h>
+
+#include <Core/PostgreSQLProtocol.h>
+#include <Formats/FormatSettings.h>
+
+namespace DB
+{
+
+class PostgreSQLOutputFormat final : public IOutputFormat
```
Comment (vitlibar) [inline src/Processors/Formats/Impl/PostgreSQLOutputFormat.h:12]: Could you please add a comment here with a link to the description of this format?

File: programs/server/PostgreSQLHandler.h (hunk 1 above)
Comment (MovElb) [reply programs/server/PostgreSQLHandler.h:20]: Ok, will be fixed.

File: programs/server/PostgreSQLHandler.cpp (hunk 2 above)
Comment (MovElb) [reply programs/server/PostgreSQLHandler.cpp:94]: It is SQL state code https://www.postgresql.org/docs/current/errcodes-appendix.html

File: tests/integration/test_postgresql_protocol/configs/users.xml (hunk 3 above)
Comment (MovElb) [reply tests/integration/test_postgresql_protocol/configs/users.xml:11]: Will be fixed.

File: src/Processors/Formats/Impl/PostgreSQLOutputFormat.h (hunk 4 above)
Comment (MovElb) [reply src/Processors/Formats/Impl/PostgreSQLOutputFormat.h:12]: Ok, I will add a comment in code. As for know you can read about it in https://www.postgresql.org/docs/current/protocol-flow.html#id-1.10.5.7.4

File: tests/integration/test_postgresql_protocol/configs/config.xml [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+     <logger>
+         <level>trace</level>
+         <log>/var/log/clickhouse-server/clickhouse-server.log</log>
+         <errorlog>/var/log/clickhouse-server/clickhouse-server.err.log</errorlog>
+         <size>1000M</size>
+         <count>10</count>
+     </logger>
+
+     <!-- Used with https_port and tcp_port_secure. Full ssl options list: https://github.com/ClickHouse-Extras/poco/blob/master/NetSSL_OpenSSL/include/Poco/Net/SSLManager.h#L71 -->
+     <openSSL>
```
Comment (vitlibar) [inline tests/integration/test_postgresql_protocol/configs/config.xml:12]: ClickHouse can be built without SSL support (i.e. with `ENABLE_SSL=0`). Is PostgreSQL protocol able to work in such build?

File: tests/integration/test_postgresql_protocol/configs/config.xml [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+             <verificationMode>none</verificationMode>
+             <loadDefaultCAFile>true</loadDefaultCAFile>
+             <cacheSessions>true</cacheSessions>
+             <disableProtocols>sslv2,sslv3</disableProtocols>
+             <preferServerCiphers>true</preferServerCiphers>
+         </server>
+     </openSSL>
+
+     <tcp_port>9000</tcp_port>
+     <postgresql_port>5433</postgresql_port>
```
Comment (vitlibar) [inline tests/integration/test_postgresql_protocol/configs/config.xml:28]: You need only `postgresql_port` to be set, do you?
Can you please remove the settings from this config file which are not necessary to set?
Just to make the test simpler.

File: src/IO/WriteHelpers.h [hunk 7]
```diff
@@ -998,4 +998,24 @@ inline String toString(const T & x)
...
 
+inline void writeNullTerminatedString(const String & s, WriteBuffer & buffer)
+{
+    buffer.write(s.data(), s.size());
+    buffer.write(0);
+}
+
+template <typename T>
+inline std::enable_if_t<is_arithmetic_v<T> && (sizeof(T) <= 8), void>
+writeBinaryBigEndian(T x, WriteBuffer & buf)    /// Assuming little endian architecture.
```
Comment (vitlibar) [inline src/IO/WriteHelpers.h:1009]: `boost::endian::native_to_big()` looks better and it requires no such assumptions.

**Comment by vitlibar** [comment]
Please fix compilation errors with `ENABLE_SSL=0`.
That must fix Yandex synchronization check.

File: tests/integration/test_postgresql_protocol/configs/config.xml (hunk 5 above)
Comment (MovElb) [reply tests/integration/test_postgresql_protocol/configs/config.xml:12]: Yes, it should work without SSL. I will check build without it.

File: tests/integration/test_postgresql_protocol/configs/config.xml (hunk 6 above)
Comment (MovElb) [reply tests/integration/test_postgresql_protocol/configs/config.xml:28]: Ok.

File: src/IO/WriteHelpers.h (hunk 7 above)
Comment (MovElb) [reply src/IO/WriteHelpers.h:1009]: Ok.

File: tests/integration/test_postgresql_protocol/configs/config.xml (hunk 6 above)
Comment (MovElb) [reply tests/integration/test_postgresql_protocol/configs/config.xml:27]: Without this option container times out, so I need to keep this option.

**Comment by alexey-milovidov** [comment]
Notes about what to do after merge:

StartUp -> Startup
MySQLWrite -> MySQLWire
PostgreSQLWrite -> PostgreSQLWire

File: src/IO/WriteHelpers.h [hunk 8]
```diff
@@ -998,4 +998,24 @@ inline String toString(const T & x)
     return buf.str();
 }
 
+inline void writeNullTerminatedString(const String & s, WriteBuffer & buffer)
+{
+    buffer.write(s.data(), s.size());
+    buffer.write(0);
```
Comment (alexey-milovidov) [inline src/IO/WriteHelpers.h:1004]: BTW, std::string is guaranteed to have terminating zero byte after its size.

File: src/IO/WriteHelpers.h (hunk 7 above)
Comment (alexey-milovidov) [reply src/IO/WriteHelpers.h:1009]: But it's Ok nevertheless :)

File: src/Processors/Formats/Impl/PostgreSQLOutputFormat.cpp [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+        for (size_t j = 0; j != chunk.getNumColumns(); ++j)
+        {
+            if (columns[j]->isNullAt(i))
+                row.push_back(std::make_shared<PostgreSQLProtocol::Messaging::NullField>());
+            else
+            {
+                WriteBufferFromOwnString ostr;
+                data_types[j]->serializeAsText(*columns[j], i, ostr, format_settings);
+                row.push_back(std::make_shared<PostgreSQLProtocol::Messaging::StringField>(std::move(ostr.str())));
```
Comment (alexey-milovidov) [inline src/Processors/Formats/Impl/PostgreSQLOutputFormat.cpp:55]: Looks slightly inefficient. Maybe zero-copy messaging is possible?
But we can merge nevertheless.

**Comment by alexey-milovidov** [comment]
The code does not have enough comments for methods and classes.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
No meaningful message is produces if there is no user.
https://youtu.be/Y7yhgjESF8s?t=8312

**Pull request merged**